include/zephyr/types.h

```diff
@@ -48,11 +48,23 @@ typedef union {
 #endif /* Z_THREAD_LOCAL */
 
 #ifdef __cplusplus
-/* Zephyr requires an int main(void) signature with C linkage for the application main if present */
+
+#if (!__STDC_HOSTED__)
+/*
+ * Zephyr requires an int main(void) signature with C linkage for the
+ * application main if present for freestanding execution.
+ *
+ * However, main cannot exist inside extern "C" linkage declarations,
+ * as that is forbidden by the C++ standard.
+ * See Section 6.9.3.1 of ISO/IEC 14882:2024.
+ *
+ * This restriction is relaxed for freestanding code, as main
+ * is not treated specially in these circumstances, so only include it
+ * when not hosted.
```

Comment:
Maybe it is better to be a bit more verbose so it is clearer?
```suggestion
 * Zephyr requires an int main(void) signature with C linkage for the application main if present.
 * gcc, and clang when building in 'hosted' mode will correctly assume this.
 * But, when building freestanding, clang does not treat main() specially, and by default name
 * mangles its symbol, which results in the linker not linking from the kernel init code into this
 * name mangled app main().
 * 
 * At the same time, according to the C++ standard Section 6.9.3.1 of ISO/IEC 14882:2024, main
 * cannot be explicitly declared to have "C" linkage.
 * This restriction is relaxed for freestanding code, as main is not treated specially in these
 * circumstances.
 * Therefore, let's include the prototype when not we are building the code as freestanding/not-
 * hosted.
```

Reply:
Reads fine; happy to make the change. I'm away from my machine right now, but should get to it later today.

Reply:
Done. Wrapped to 80 chars like the other comments in this file.